core/src/main/java/org/apache/calcite/rel/metadata/RelMdUtil.java

```diff
@@ -210,6 +210,18 @@ public static boolean areColumnsDefinitelyUnique(RelMetadataQuery mq,
     return b != null && b;
   }
 
+  public static boolean isRelDefinitelyEmpty(RelMetadataQuery mq,
+      RelNode rel) {
+    Boolean b = mq.isEmpty(rel);
+    return b != null && b;
+  }
+
+  public static boolean isRelDefinitelyNotEmpty(RelMetadataQuery mq,
```

Comment:
Can this method be replaced by "!isRelDefinitelyEmpty"?

Reply:
It can't because the isEmpty value can be null.